Bluetooth: Mesh: Mesh Settings commit should after BT

LingaoM opened the pull request:
Mesh Settings Commit should after BT.

This solves the mesh stack (and actually any module) dependency on the host being able to register a GATT service:
https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/subsys/bluetooth/host/gatt.c#L1737-L1742

which (the GATT service registration) is not allowed until `sc_commit` sets `SC_LOAD` flag.
Which in turns requires the mesh stack (which is removed in this PR) to re-schedule an additional delayable work item to try register GATT service over and over again until `sc_commit` in `subsys/bluetooth/host/gatt.c` is called.

So it is not about loading. But as @LingaoM said it is about order of `h_commit` calls for Bluetooth subsystem.

```
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(._settings_handler_static.static.*)))
 ._settings_handler_static.static.settings_handler_0_bt_
                0x0000000000034644       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(settings.c.obj)
                0x0000000000034644                settings_handler_0_bt
 ._settings_handler_static.static.settings_handler_0_bt_ccc_
                0x0000000000034658       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034658                settings_handler_0_bt_ccc
 ._settings_handler_static.static.settings_handler_0_bt_cf_
                0x000000000003466c       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x000000000003466c                settings_handler_0_bt_cf
 ._settings_handler_static.static.settings_handler_0_bt_hash_
                0x0000000000034680       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034680                settings_handler_0_bt_hash
 ._settings_handler_static.static.settings_handler_0_bt_sc_
                0x0000000000034694       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034694                settings_handler_0_bt_sc
 ._settings_handler_static.static.settings_handler_1_bt_mesh_
                0x00000000000346a8       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(settings.c.obj)
                0x00000000000346a8                settings_handler_1_bt_mesh
 ._settings_handler_static.static.settings_handler_1_bt_mesh_app_
                0x00000000000346bc       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(app_keys.c.obj)
                0x00000000000346bc                settings_handler_1_bt_mesh_app
 ._settings_handler_static.static.settings_handler_1_bt_mesh_cfg_
                0x00000000000346d0       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(cfg.c.obj)
                0x00000000000346d0                settings_handler_1_bt_mesh_cfg
 ._settings_handler_static.static.settings_handler_1_bt_mesh_comp_
                0x00000000000346e4       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(access.c.obj)
                0x00000000000346e4                settings_handler_1_bt_mesh_comp
 ._settings_handler_static.static.settings_handler_1_bt_mesh_iv_
                0x00000000000346f8       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(net.c.obj)
                0x00000000000346f8                settings_handler_1_bt_mesh_iv
 ._settings_handler_static.static.settings_handler_1_bt_mesh_net_
                0x000000000003470c       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(net.c.obj)
                0x000000000003470c                settings_handler_1_bt_mesh_net

```

LingaoM committed:
Bluetooth: Mesh: Mesh Settings commit should after BT

Mesh Settings Commit should after BT.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

PavelVPV commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -117,7 +117,7 @@ static int mesh_commit(void)
 }
 
-SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
+SETTINGS_STATIC_HANDLER_DEFINE(_2_bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
```
Shouldn't `BT_MESH_SETTINGS_DEFINE` use this prefix as well?

LingaoM replied:
```
#define BT_MESH_SETTINGS_DEFINE(_hname, _subtree, _set)
```
 only define set, not commit


``` C
int settings_load_subtree(const char *subtree)
{
	struct settings_store *cs;
	int rc;
	const struct settings_load_arg arg = {
		.subtree = subtree
	};

	/*
	 * for every config store
	 *    load config
	 *    apply config
	 *    commit all
	 */
	k_mutex_lock(&settings_lock, K_FOREVER);
	SYS_SLIST_FOR_EACH_CONTAINER(&settings_load_srcs, cs, cs_next) {
		cs->cs_itf->csi_load(cs, &arg);
	}
	rc = settings_commit_subtree(subtree);
	k_mutex_unlock(&settings_lock);
	return rc;
}
```

The most important thing is the order of commits

LingaoM replied:
> There are few more places where SETTINGS_STATIC_HANDLER_DEFINE is used by bluetooth related models. Shouldn't they be updated as well?
controller/ll_sw/ll_settings.c:92:SETTINGS_STATIC_HANDLER_DEFINE(bt_ctlr, "bt/ctlr", NULL, ctlr_set, NULL, NULL);
host/settings.h:16:	SETTINGS_STATIC_HANDLER_DEFINE(bt_##_hname, "bt/" _subtree, NULL, _set, _commit, NULL)
host/settings.c:284:SETTINGS_STATIC_HANDLER_DEFINE(bt, "bt", NULL, set_setting, commit_settings, NULL);
host/classic/keys_br.c:220:SETTINGS_STATIC_HANDLER_DEFINE(bt_link_key, "bt/link_key", NULL, link_key_set,
mesh/dfu_slot.c:401:SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_dfu_slots, DFU_SLOT_SETTINGS_PATH, NULL,
mesh/settings.h:39:	SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_##_hname, "bt/mesh/" _subtree, NULL, pre_##_set,    \
mesh/settings.c:120:SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
services/dis.c:453:SETTINGS_STATIC_HANDLER_DEFINE(bt_dis, "bt/dis", NULL, dis_set, NULL, NULL);

I confirmed that these are just sets, not commits.

This is also a reply to the third point above. https://github.com/zephyrproject-rtos/zephyr/pull/79690#pullrequestreview-2362040505

PavelVPV replied:
But if someone adds a commit handler for some of mesh settings, this can cause a bug, can't it?

LingaoM replied:
Before this PR, if I did what you said, there would still be a problem of status being out of sync.

LingaoM replied:
Mainly inspired by: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/gatt.c#L234

PavelVPV reviewed (commented):
Fantastic! Few suggestions for improvement:
1. I think we need a comment in front of the host and mesh settings handler definitions why it is like that, or see the next point
2. Might be a good idea to create `settings.h` file under `subsys/bluetooth/common` folder and define all these bt prefixes there.
3. There are few more places where `SETTINGS_STATIC_HANDLER_DEFINE` is used by bluetooth related models. Shouldn't they be updated as well?
```
controller/ll_sw/ll_settings.c:92:SETTINGS_STATIC_HANDLER_DEFINE(bt_ctlr, "bt/ctlr", NULL, ctlr_set, NULL, NULL);
host/settings.h:16:	SETTINGS_STATIC_HANDLER_DEFINE(bt_##_hname, "bt/" _subtree, NULL, _set, _commit, NULL)
host/settings.c:284:SETTINGS_STATIC_HANDLER_DEFINE(bt, "bt", NULL, set_setting, commit_settings, NULL);
host/classic/keys_br.c:220:SETTINGS_STATIC_HANDLER_DEFINE(bt_link_key, "bt/link_key", NULL, link_key_set,
mesh/dfu_slot.c:401:SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_dfu_slots, DFU_SLOT_SETTINGS_PATH, NULL,
mesh/settings.h:39:	SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_##_hname, "bt/mesh/" _subtree, NULL, pre_##_set,    \
mesh/settings.c:120:SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
services/dis.c:453:SETTINGS_STATIC_HANDLER_DEFINE(bt_dis, "bt/dis", NULL, dis_set, NULL, NULL);
```

LingaoM committed:
tests: bsim: fixes Failed updating adv params: -105

<err> bt_mesh_adv_ext: Failed updating adv params: -105
<wrn> bt_mesh_gatt: Failed to advertise using Network ID
(err -105)
<err> bt_mesh_gatt: Advertising proxy failed (err: -105)

advertiser/proxy_mixin.sh FAILED (1.432 s)

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

Thalley commented on subsys/bluetooth/host/settings.c:
```diff
@@ -281,7 +281,8 @@ static int commit_settings(void)
 }
 
-SETTINGS_STATIC_HANDLER_DEFINE(bt, "bt", NULL, set_setting, commit_settings, NULL);
+/* Add `_0` prefix to ensure BT always commit firstly. */
+SETTINGS_STATIC_HANDLER_DEFINE(_0_bt, "bt", NULL, set_setting, commit_settings, NULL);
```
That's not really a good solution IMO

If something is `00_` then it still comes before `0_bt`. 

Unfortunately there doesn't seem to be any maintainer for the settings subsystem, but if we want to add priorities or specific order of settings, then I think we need to consider better alternatives. 

At the top of my head, suggestions could be

1. Add a priority in the API (either the BT_SETTINGS or SETTINGS API)
2. Have a single settings handler in BT that can load the others in the right order. 

This solution here seems to use an undocumented approach that is easy to get wrong and is based on the assumption that the names will always control the order

LingaoM replied:
Done.

LingaoM replied:
``` C
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(._settings_handler_static.static.*)))
 ._settings_handler_static.static.settings_handler_0_bt_
                0x0000000000034644       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(settings.c.obj)
                0x0000000000034644                settings_handler_0_bt
 ._settings_handler_static.static.settings_handler_0_bt_ccc_
                0x0000000000034658       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034658                settings_handler_0_bt_ccc
 ._settings_handler_static.static.settings_handler_0_bt_cf_
                0x000000000003466c       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x000000000003466c                settings_handler_0_bt_cf
 ._settings_handler_static.static.settings_handler_0_bt_hash_
                0x0000000000034680       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034680                settings_handler_0_bt_hash
 ._settings_handler_static.static.settings_handler_0_bt_sc_
                0x0000000000034694       0x14 zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a(gatt.c.obj)
                0x0000000000034694                settings_handler_0_bt_sc
 ._settings_handler_static.static.settings_handler_1_bt_mesh_
                0x00000000000346a8       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(settings.c.obj)
                0x00000000000346a8                settings_handler_1_bt_mesh
 ._settings_handler_static.static.settings_handler_1_bt_mesh_app_
                0x00000000000346bc       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(app_keys.c.obj)
                0x00000000000346bc                settings_handler_1_bt_mesh_app
 ._settings_handler_static.static.settings_handler_1_bt_mesh_cfg_
                0x00000000000346d0       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(cfg.c.obj)
                0x00000000000346d0                settings_handler_1_bt_mesh_cfg
 ._settings_handler_static.static.settings_handler_1_bt_mesh_comp_
                0x00000000000346e4       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(access.c.obj)
                0x00000000000346e4                settings_handler_1_bt_mesh_comp
 ._settings_handler_static.static.settings_handler_1_bt_mesh_iv_
                0x00000000000346f8       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(net.c.obj)
                0x00000000000346f8                settings_handler_1_bt_mesh_iv
 ._settings_handler_static.static.settings_handler_1_bt_mesh_net_
                0x000000000003470c       0x14 zephyr/subsys/bluetooth/mesh/libsubsys__bluetooth__mesh.a(net.c.obj)
                0x000000000003470c                settings_handler_1_bt_mesh_net
```

Laczen commented:
I don't exactly understand what is being targeted here. It seems you need some kind of order in the processing of settings? When loading settings they will come in the order they are found on storage, there is no ordering here.

PavelVPV commented:
> I don't exactly understand what is being targeted here. It seems you need some kind of order in the processing of settings? When loading settings they will come in the order they are found on storage, there is no ordering here.

This solves the mesh stack (and actually any module) dependency on the host being able to register a GATT service:
https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/subsys/bluetooth/host/gatt.c#L1737-L1742

which (the GATT service registration) is not allowed until `sc_commit` sets `SC_LOAD` flag.
Which in turns requires the mesh stack (which is removed in this PR) to re-schedule an additional delayable work item to try register GATT service over and over again until `sc_commit` in `subsys/bluetooth/host/gatt.c` is called.

So it is not about loading. But as @LingaoM said it is about order of `h_commit` calls for Bluetooth subsystem.

alxelax commented:
> > I don't exactly understand what is being targeted here. It seems you need some kind of order in the processing of settings? When loading settings they will come in the order they are found on storage, there is no ordering here.
> 
> This solves the mesh stack (and actually any module) dependency on the host being able to register a GATT service:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/subsys/bluetooth/host/gatt.c#L1737-L1742
> 
> which (the GATT service registration) is not allowed until `sc_commit` sets `SC_LOAD` flag. Which in turns requires the mesh stack (which is removed in this PR) to re-schedule an additional delayable work item to try register GATT service over and over again until `sc_commit` in `subsys/bluetooth/host/gatt.c` is called.
> 
> So it is not about loading. But as @LingaoM said it is about order of `h_commit` calls for Bluetooth subsystem.

Isn't this question of subtree loading sequence? This is sample how I solved the same issue for bsim tests when it emulates secure storage:
https://github.com/zephyrproject-rtos/zephyr/commit/8a51176877c9b74e286303a452c3b5d316d90e96#diff-a075de11be69d631fe1e1944a1782d1c7665a6f8b2ae02be1e355e1de2682562R257-R262

PavelVPV commented:
> > > I don't exactly understand what is being targeted here. It seems you need some kind of order in the processing of settings? When loading settings they will come in the order they are found on storage, there is no ordering here.
> > 
> > This solves the mesh stack (and actually any module) dependency on the host being able to register a GATT service:
> > 
> > https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/subsys/bluetooth/host/gatt.c#L1737-L1742
> > 
> > which (the GATT service registration) is not allowed until `sc_commit` sets `SC_LOAD` flag. Which in turns requires the mesh stack (which is removed in this PR) to re-schedule an additional delayable work item to try register GATT service over and over again until `sc_commit` in `subsys/bluetooth/host/gatt.c` is called.
> > 
> > So it is not about loading. But as @LingaoM said it is about order of `h_commit` calls for Bluetooth subsystem.
> 
> Isn't this question of subtree loading sequence? This is sample how I solved the same issue for bsim tests when it emulates secure storage:
> https://github.com/zephyrproject-rtos/zephyr/commit/8a51176877c9b74e286303a452c3b5d316d90e96#diff-a075de11be69d631fe1e1944a1782d1c7665a6f8b2ae02be1e355e1de2682562R257-R262
> 

The suggestion from this commit requires changing all existing applications. Moreover, it brings internal information (subtree paths) to the applications. Also it does not allow the use of `settings_load()`.

This PR does not require any changes in the existing apps.

PavelVPV commented:
> Isn't this question of subtree loading sequence?

No, loading order doesn't matter. Service registration happens in the commit phase (when `h_commit` is called).

alxelax commented:
> h_commit

I do not think you are right. Please take a look:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/settings/src/settings_store.c#L60

@Laczen, I guess, it is still possible to solve by slightly modifying `subtree_load` function for bluetooth purposes. When it goes over subtrees in loop, it will ignore bluetooth related subtrees and load them separately in predefined order.

@PavelVPV, it will not require changing samples, applications etc just `settings.c`. It will be still possible to use `settings_load()` as is.

PavelVPV commented:
> OK, the proposal with `bt_init_levels` also handles the order of `h_commit` calls. It seems fragile to depend on alphabetical ordering of linker sections (and that static handlers are processed before dynamic handlers).


The order is guaranteed by ITERABLE_SECTION_ROM macro:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include%2Fzephyr%2Flinker%2Fiterable_sections.h#L38-L51

alxelax commented:
> > OK, the proposal with `bt_init_levels` also handles the order of `h_commit` calls. It seems fragile to depend on alphabetical ordering of linker sections (and that static handlers are processed before dynamic handlers).
> 
> The order is guaranteed by ITERABLE_SECTION_ROM macro: https://github.com/zephyrproject-rtos/zephyr/blob/main/include%2Fzephyr%2Flinker%2Fiterable_sections.h#L38-L51

Order is guaranteed by this loop first of all:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/settings/src/settings_store.c#L57-L59

PavelVPV commented:
> Order of loading and commitment is related to each other. 

This PR is not about that, but about the order of h_commit calls.

alxelax commented:
> > Order of loading and commitment is related to each other.
> 
> This PR is not about that, but about the order of h_commit calls.

Probably this is my fault since I cannot deliver idea about that the order of `h_commit` depends on order of loaded subtrees.
Even this PR confirms it. It just sorts subtrees in alphabetical order that impacts on order of loading and as consequence on order of `h_commit` calls.

alxelax reviewed (changes requested):
but agree with @Laczen and @Thalley this is a fragile solution, that uses hidden knowledge and doesn't seem acceptable.

Laczen commented:
@PavelVPV, maybe we can put the return value of the `h_commit` into use and restart the commit loop if an `ECANCELED` is returned. We should of course limit the amount of restarts.

PavelVPV commented:
> @PavelVPV, maybe we can put the return value of the `h_commit` into use and restart the commit loop if an `ECANCELED` is returned. We should of course limit the amount of restarts.

Making an assumption in the mesh stack that the host didn't allow registering a GATT service because _probably_ it didn't complete its initialization is complicated and risky. We can't rollback what the mesh stack already initialized. The host doesn't also differentiate this error with other errors and returns `-EINVAL` (`gatt_register` also returns `-EINVAL`).

PavelVPV commented:
> @PavelVPV, maybe we can put the return value of the `h_commit` into use and restart the commit loop if an `ECANCELED` is returned. We should of course limit the amount of restarts.

Since this is a bluetooth subsystem specific issue, the last [commit](https://github.com/zephyrproject-rtos/zephyr/pull/79690/commits/acfb64f9a624a55bd9e6924adfd308f9042a1df6) with `BT_SETTINGS_PRIO_BT` and `BT_SETTINGS_PRIO_AFT_BT` looks good to me.

PavelVPV commented:
We can document in settings that static settings handlers are commited in the alphabetical order since settings uses `ITERABLE_SECTION_ROM` macro which guarantees the order.

LingaoM commented:
> We can document in settings that static settings handlers are commited in the alphabetical order since settings uses `ITERABLE_SECTION_ROM` macro which guarantees the order.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/linker/iterable_sections.h#L17
Following byte order is already the default behavior of the system.

Laczen commented:
@LingaoM, @PavelVPV, I think that what is proposed here could be added and will solve the problem you are facing. 

However I think it is a sign of a more general problem that appears in settings. Sometimes commits cannot be handled directly and some commit should be postponed. This is solvable by the technique proposed when it appears in a subsystem but cannot be handled if there is a "dependency" between multiple subsystems.

Therefore it would be good if the settings subsystem would either restart the commit list or a "unhandled" commit list is created that will be handled in a second stage. This could be updated and (boundedly) restarted.

I would go for the presently proposed solution but create an issue that the settings system needs to be modified to handle the more general problem.

alxelax commented:
> We can document in settings that static settings handlers are commited in the alphabetical order since settings uses `ITERABLE_SECTION_ROM` macro which guarantees the order.

Until it is not documented and well described in public settings API this PR is implementation based on internal implementation third subsytem.

PavelVPV commented:
> @LingaoM, @PavelVPV, I think that what is proposed here could be added and will solve the problem you are facing.
> 
> However I think it is a sign of a more general problem that appears in settings. Sometimes commits cannot be handled directly and some commit should be postponed. This is solvable by the technique proposed when it appears in a subsystem but cannot be handled if there is a "dependency" between multiple subsystems.
> 
> Therefore it would be good if the settings subsystem would either restart the commit list or a "unhandled" commit list is created that will be handled in a second stage. This could be updated and (boundedly) restarted.
> 
> I would go for the presently proposed solution but create an issue that the settings system needs to be modified to handle the more general problem.

Some subsystems may then need 2 - 3 iterations (depending on subsystems they depend on) before they could complete the commit routine. In that sense I'd rather keep using `k_work_delayable` with pooling like it has been done and hope it won't fail. The amount of logic we need to build on top of the mesh settings' `h_commit` handler doesn't worth the effort. The approach with `k_work_delayable` can be used by anyone without additional overhead in settings or any other subsystem they depend on. I have a feeling this can be more bug prone.

I still think the proposed solution is good enough as it doesn't add any overhead to any of the current subsystems. It also makes sense for me that the Bluetooth host completes its commit routine and the rest Bluetooth subsystems do this after the host.

The order of `h_commit` calls for statically defined settings handler has been alphabetical de-facto and I bet it would not be a problem if the mesh `h_commit` handler by coincidence would be called after the host's `sc_commit`.

LingaoM commented:
> > @LingaoM, @PavelVPV, I think that what is proposed here could be added and will solve the problem you are facing.
> > However I think it is a sign of a more general problem that appears in settings. Sometimes commits cannot be handled directly and some commit should be postponed. This is solvable by the technique proposed when it appears in a subsystem but cannot be handled if there is a "dependency" between multiple subsystems.
> > Therefore it would be good if the settings subsystem would either restart the commit list or a "unhandled" commit list is created that will be handled in a second stage. This could be updated and (boundedly) restarted.
> > I would go for the presently proposed solution but create an issue that the settings system needs to be modified to handle the more general problem.
> 
> Some subsystems may then need 2 - 3 iterations (depending on subsystems they depend on) before they could complete the commit routine. I have a feeling this can be more bug prone. In that sense I'd rather keep using `k_work_delayable` with pooling like it has been done and hope it won't fail. The amount of logic we need to build on top of the mesh settings' `h_commit` handler doesn't worth the effort. The approach with `k_work_delayable` can be used by anyone without additional overhead in settings or any other subsystem they depend on.
> 
> I still think the proposed solution is good enough as it doesn't add any overhead to any of the current subsystems. It also makes sense for me that the Bluetooth host completes its commit routine first and the rest Bluetooth subsystems do this after the host.
> 
> The order of `h_commit` calls for statically defined settings handler has been alphabetical de-facto and I bet it would not be a problem if the mesh `h_commit` handler by coincidence would be called after the host's `sc_commit`.

Fully Agree.

Laczen commented:
@LingaoM , @PavelVPV as I have repeatedly stated I would agree with the solution as I think it is good enough for the present problem.

However I do think a follow-up is needed: the approach with `k_work_delayable` does not seem a sufficient solution or we would not be discussing this.

I am not at all pursuing the need to change anything in the bluetooth subsystem at the moment, whatever solution is proposed it should not require direct changes, but only enable a more generic solution. I have been thinking about how to solve the general problem and I would do it as:
1. Introduce a settings Kconfig that sets a limit on the commit reprocessing of handlers (setting this to 0 should have 0 effect on code size).
2. When the reprocessing is set to a value different from 0 each `h_commit` can return a `-ECANCELLED` that results in the handler being placed in a `cancelled_handlers` slist,
3. When the `static handlers` are processed the `cancelled_handlers` slist is processed and every succesfull `h_commit` removes it from the `cancelled_handlers` slist.

This should ensure:
1. Each handler is only called once with succesfull result,
2. No code is added when the commit reprocessing count is 0,
3. Allow the proposed change to work without any required extra changes to the bt subsystem, but enabling future changes (if required/deemed advantageous).

This might retrigger going over the `cancelled_handlers` slist, but this shouldn't be a problem as long as it is bounded.

@alxelax, @Thalley, could you remove the blocking change request ? A follow-up should be capable of providing a more general correction.

Thalley commented:
Before I remove my change request, I need to fully understand these new changes. To me it looks like the priority is implemented using the name of the settings. 

Is that correct? 
If so, can we rely on this for only the Zephyr SDK, or any compilers/linkers?
If so, would the change in this PR effectively remove (by not reading existing) settings for all current devices?

Laczen commented:
> Before I remove my change request, I need to fully understand these new changes. To me it looks like the priority is implemented using the name of the settings.
> 
> Is that correct? If so, can we rely on this for only the Zephyr SDK, or any compilers/linkers? If so, would the change in this PR effectively remove (by not reading existing) settings for all current devices?

The settings names are left unchanged, the settings_handler names are changed, so there should be no problem for current devices.

PavelVPV commented:
@Laczen, I'm not against what you proposal, but I don't see that it can be easily applied in our case.

This is the call stack for the Mesh Proxy GATT service registration which happens during `settings_load()` without `k_work_delayable` (like in this PR):

[mesh_commit](https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/subsys/bluetooth/mesh/settings.c#L85-L115) -> [bt_mesh_start](https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/subsys/bluetooth/mesh/main.c#L622-L645) -> [bt_mesh_proxy_gatt_enable](https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/subsys/bluetooth/mesh/proxy_srv.c#L975-L997) -> [bt_gatt_service_register](https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/subsys/bluetooth/host/gatt.c#L1729-L1742)
 
If we want to make use of the solution you propose, we need:
- make `bt_gatt_service_register` return an error different than `-EINVAL` if the service registration fails due to the host not being ready yet. If this is considered as an API change (a new error code for old behavior), we need to pass the bt wg and architecture reviews.
- propagate the error from `bt_mesh_proxy_gatt_enable` up to `mesh_commit`. Which means that `bt_mesh_start` should then return `-ECANCELED` or any other meaningful error that can be converted to `-ECANCELED` in `mesh_commit` function. This IMHO is very confusing if a function completes its job but returns an error code. 
- add a flag to `bt_mesh.flags` that the service registration failed.
- check this flag in `mesh_commit` and explicitly call `bt_mesh_proxy_gatt_enable` one more time.

An option could be to add an API function to the host which the mesh can call from `mesh_commit` to check if the host is ready to register a GATT service. But this anyway adds more overhead than the solution from the PR. We should also make sure that the Kconfig option (for reprocessing the commit handlers) is correctly configured. From what I also see all bluetooth samples ignore `settings_load` error code. So if `settings_load` fails due to this, users may not see why mesh is not running.

Also, if there is a dependency chain, the configuration of the Kconfig option (for reprocessing the commit handlers) (I think) becames complicated as sub-dependencies should also be considered. This may eventually involve a user for proper configuration.

I may of course be wrong and don't see the whole picture so appologize in advance.

Laczen commented:
> @Laczen, I'm not against what you proposal, but I don't see that it can be easily applied in our case.
>
That is OK, as said there is no intention to enforce my proposal to the bt subsystem. It would just be a follow-up, if the bt subsystem can solve it without my proposal this would be excellent. Even more, all functionality that is available today should remain (including the present PR).

> Also, if there is a dependency chain, the configuration of the Kconfig option (for reprocessing the commit handlers) (I think) becames complicated as sub-dependencies should also be considered. This may eventually involve a user for proper configuration.
>
The Kconfig option would only be to enable reprocessing and setting a limit to the reprocessing. And you are correct, my proposal might not solve all situations, it is possible to create a dependency that will never be solved by reprocessing the handlers.

Maybe an even simpler solution than my proposal is the proposal of @Thalley to include a commit priority in the settings handlers and to have subsystems use this to set a commit priority. This would be a very small change.

PavelVPV commented:
> Maybe an even simpler solution than my proposal is the proposal of @Thalley to include a commit priority in the settings handlers and to have subsystems use this to set a commit priority. This would be a very small change.

I'd say this is already solved by sorting static handlers. Dynamic handlers can be sorted at the registration.

alxelax commented:
> @alxelax, @Thalley, could you remove the blocking change request ? A follow-up should be capable of providing a more general correction.

Hi @Laczen, I will as soon as feature with ordered static settings entries will be documented and became public documented feature of the settings subsystem.
At the current moment, this PR is based on internal implementation on another module. That might be changed. It for instance has happened recently in Bluetooth Host where adv tx queue handling were redesigned. Some assumptions about another module behavior became not working that caused significant problems. Adding more “assumption based” solution shouldn't be done, I'm quite sure.

PavelVPV commented:
@LingaoM , could you please update settings documentation saying that order of `h_commit` calls for settings handlers initialized by `SETTINGS_STATIC_HANDLER_DEFINE` is alphabetical? I could also do it, but much later.

PavelVPV commented on subsys/bluetooth/common/settings.h:
Would be nice to see description for the new macros.

LingaoM replied:
BTW: bbsim always failed, i'm confuse that, just re-order _commit, have influence of bbsim.

PavelVPV replied:
Will you investigate this? I need to prioritize other work now.

PavelVPV committed:
bluteooth: mesh: Start scanner before advertiser

This fixes legacy advertiser issue. When Starting or resuming mesh,
advertiser thread is started before scanner. If it happens that we
enable scanner while initiated advertising, the controller won't
allow to start scanner.

This is a dirty fix to start scanner first as the scanner is started
directly in `bt_mesh_start` and `bt_mesh_resume`, while advertising in
case of the legacy advertiser is started in a separate thread.

This may affect extended advertiser as well.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
bluetooth: mesh: Fix recursive advertsier start

Using `bt_mesh_adv_gatt_start` outside adv_ext.c is an issue because
when it is used on `bt_mesh_resume`, the `ADV_FLAG_PROXY` is not set,
and the mesh extended advertiser doesn't know that proxy advs is
started. Thus it will try to start it one more time, but controller will
reject this.

I think we should not use `bt_mesh_adv_gatt_start` outside the mesh
advertiser and use `bt_mesh_adv_gatt_update` instead.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bsim: bluetooth: mesh: Add delay for tx test

This delay shifts transmission of advs in the tx device. After the
GATT service registration move from workq, the service is registerd
earlier and tx device starts advertising earlier.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
@LingaoM ,

The `tests_scripts/suspend/gatt_suspend_resume.sh` test is failing because of recursive `bt_mesh_adv_gatt_start` call, which was happening because we called `bt_mesh_adv_gatt_send` outside `adv_ext.c` without setting `ADV_FLAG_PROXY` flag (like we do in `adv_ext.c` in `send_pending_adv` handler).
I've created a temporary fix, but not fully sure if it is correct: https://github.com/zephyrproject-rtos/zephyr/commit/e8fd5d4342c3b554037c733f4abaa70e8fc1299e

The `tests_scripts/suspend/gatt_suspend_disable_resume.sh` test is failing in the legacy advertiser because we were trying to start scanner while advertising and the controller didn't allow us to do that. I've created a very dirty fix by reordering scanner start and advertiser start, but we should think more how to fix this correctly: https://github.com/zephyrproject-rtos/zephyr/commit/f0e1506b2af3582f53eafde22b2007c345650432

The `tests_scripts/priv_beacon/proxy_adv_multi_subnet_coex.sh` test is failing simply because of timings change after registering GATT service right in the `bt_mesh_start` call instead of in work handler later. I've added a delay, but the test should be properly fixed: https://github.com/zephyrproject-rtos/zephyr/commit/90de90c4f8fdc1fd4417192e104b282fa56bf912

I would really appreciate if you fix all these issues since you started this PR. I don't have time atm to fix them. But the first two are quite critical I think.

Thalley commented:
Please also have a look at https://github.com/zephyrproject-rtos/zephyr/pull/79855 from @Laczen :)

PavelVPV commented:
> Please also have a look at #79855 from @Laczen :)

Make sense. It adds just a little overhead with `cprio` field, but overall this looks good. Though something like `subsys/bluetooth/common/settings.h` will still be required to run host's h_commit before mesh one.

PavelVPV commented:
> I’m curious, is this PR still worth merging after #79855?, or replace `SETTINGS_STATIC_HANDLER_DEFINE` in common.h with `SETTINGS_STATIC_HANDLER_DEFINE_WITH_CPRIO` based on this PR(#79855).

The second option with `cprio`.

Laczen commented:
> I’m curious, is this PR still worth merging after #79855?, or replace `SETTINGS_STATIC_HANDLER_DEFINE` in common.h with `SETTINGS_STATIC_HANDLER_DEFINE_WITH_CPRIO` based on this PR(#79855).

@LingaoM It was not my intention for #79855 to intervene with the present PR. The present PR needed a follow-up to ensure that other users would not need to rely on alphabetic ordering (that would be hard for dynamic handlers), #79855 is the follow-up. IMHO #79855 can probably be used to simplify some of the bluetooth setup, and I hope you can use it to your advantage.

LingaoM commented:
> > I’m curious, is this PR still worth merging after #79855?, or replace `SETTINGS_STATIC_HANDLER_DEFINE` in common.h with `SETTINGS_STATIC_HANDLER_DEFINE_WITH_CPRIO` based on this PR(#79855).
> 
> @LingaoM It was not my intention for #79855 to intervene with the present PR. The present PR needed a follow-up to ensure that other users would not need to rely on alphabetic ordering (that would be hard for dynamic handlers), #79855 is the follow-up. IMHO #79855 can probably be used to simplify some of the bluetooth setup, and I hope you can use it to your advantage.

The original intention of this PR is to solve the order of commits, thereby avoiding the logic of mesh retry. Generally speaking, this purpose has been achieved :).
